feat: Enforce 2FA for members of an organisation

jemiluv8 opened the pull request:
## What does this PR do?
Introduce another field in the organization.metadata JSON field to use to configure whether organizations require 2fa for their members. 

Introduce a security section under organisation section of sidebar where an organisation's admin can update 
the above field to determine whether 2fa ought to be required on all organization members.

<img width="1438" alt="Screenshot 2023-08-17 at 11 31 31 AM" src="https://github.com/calcom/cal.com/assets/119384208/5a5c7a22-f859-40dc-b999-279d3ee9c4ed">

Enforces 2FA for users belonging to an organization by redirecting users that have not setup mfa
to the /settings/security/two-factor-auth page. 
A banner is also shown to indicate the 2fa is required on the account
The user is prevented from accessing any other pages until 2fa is setup

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Fixes #8432
/claim #8432

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->
https://www.loom.com/share/73c9c35e80d14be7a9ad33cb50ca2d3b?sid=ba232f65-364f-41bb-8ca1-854e344f88fe

## Requirement/Documentation

## Type of change

<!-- Please delete bullets that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?
- [X] Login with a user that belongs to an organization but has not setup 2fa
- [X] Verify that you're redirected to the relevant screen to setup 2fa
- [X] Verify that you cannot access any other screen until 2fa is setup 
- [X] Verify that you can access other pages after 2fa is setup and the banner disappears
- [X] Verify also that this change does not affect "non-organizational" users

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->

- ~I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)~
- ~My code doesn't follow the style guidelines of this project~
- ~I haven't commented my code, particularly in hard-to-understand areas~
- ~I haven't checked if my PR needs changes to the documentation~
- ~I haven't checked if my changes generate no new warnings~
- ~I haven't added tests that prove my fix is effective or that my feature works~
- ~I haven't checked if new and existing unit tests pass locally with my changes~

jemiluv8 committed:
Enforce 2FA on accounts belonging to active team members of organizations

We just redirect to the /settings/security/two-factor-auth everytime a logged in user attempts to go elsewhere

jemiluv8 committed:
enforce 2fa for organization member

jemiluv8 commented:
@leog, I’ve had some challenge testing this PR against “actual” members of an organisation.
I’m not sure how I can create an organization. Can you give me some pointers?

I'm currently following the README in the packages/features/ee folder to enable SAML
login.

jemiluv8 commented on apps/web/pages/index.tsx:
```diff
@@ -13,6 +13,13 @@ export async function getServerSideProps({ req, res }: GetServerSidePropsContext
   }
 
+  if (session.user) {
+    const { organizationId = null, twoFactorEnabled = false } = session.user;
```
Ensure user is redirected to setup mfa if they belong to an organization

jemiluv8 commented on packages/features/auth/lib/getServerSession.ts:
```diff
@@ -73,6 +73,7 @@ export async function getServerSession(options: {
       impersonatedByUID: token.impersonatedByUID ?? undefined,
       belongsToActiveTeam: token.belongsToActiveTeam,
       organizationId: token.organizationId,
+      twoFactorEnabled: user.twoFactorEnabled,
```
added ```twoFactorEnabled``` to session data to help us enforce 2fa on organization memebers.

jemiluv8 commented on packages/features/shell/Shell.tsx:
```diff
@@ -189,6 +191,28 @@ function useRedirectToOnboardingIfNeeded() {
   };
 }
 
+function useRedirectTo2FAIfNeeded() {
```
hook to redirect organiztion user to two-factor-auth web page whenever they are navigating within the app

jemiluv8 commented on packages/features/users/components/TwoFactorAuthRequiredBanner.tsx:
```diff
@@ -0,0 +1,24 @@
+import { TopBanner } from "@calcom/ui";
+import { Key } from "@calcom/ui/components/icon";
+
+function TwoFactorAuthRequiredBanner() {
```
Banner to prompt users that they are required to setup mfa.
This helps avoid confusion since they will not be able to navigate
to other pages until they setup 2fa.

jemiluv8 reviewed (commented):
Self review completed

jemiluv8 committed:
useMeQuery instead of session to fetch user data with info regarding 2fa

jemiluv8 committed:
remove comment

jemiluv8 committed:
introduce settings for organizations to enforce twoFactor

jemiluv8 committed:
enforce 2fa only if required by organization

jemiluv8 committed:
add requireTwoFactor to User on Session
do some refactoring

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=10584) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=10584) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=10584) it.</sub>

jemiluv8 committed:
Merge branch 'main' of https://github.com/jemiluv8/cal.com into enforce-2fa-for-org-members

jemiluv8 committed:
Change implementation of two factor auth enforcement

- store org configuration for requiring two factor auth on all members in
   organization's metadata instead of as a field

jemiluv8 commented on packages/trpc/server/routers/viewer/organizations/create.handler.ts:
```diff
@@ -120,7 +120,7 @@ export const createHandler = async ({ input, ctx }: CreateOptions) => {
         });
       } else {
-        console.warn("Organization created: subdomain not configured and couldn't notify adminnistrators");
+        console.warn("Organization created: subdomain not configured and couldn't notify administrators");
```
Fixed typo

icsy7867 replied:
Would love to have this feature for our organization.

jemiluv8 committed:
fix typo and remove redundant console logs

jemiluv8 committed:
Merge branch 'main' into enforce-2fa-for-org-members

jemiluv8 committed:
update implementation to use new org field on session user

jemiluv8 commented on package.json:
```diff
@@ -107,8 +107,7 @@
     "(apps|packages)/**/*.{js,ts,jsx,tsx}": [
-      "prettier --write",
-      "eslint --fix"
+      "prettier --write"
```
I couldn't commit because eslint --fix was failing because of an error that had nothing to do with my changes.

jemiluv8 committed:
remove unused variable and join on organization

jemiluv8 committed:
fix type error

jemiluv8 committed:
fix more type errors

jemiluv8 committed:
Merge branch 'main' into enforce-2fa-for-org-members

jemiluv8 commented:
still open
